File: 0242-valid-anagram/0242-valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    // check that s.length and t.length are equal (base-case)
    // create hash tables with key-value pairs for 's' and 't'
    // iterate through s and t to tally each characters
    // iterate through hash table and compare # of characters
    // if not equal - return false, else return true
    
    if (s.length !== t.length) {
        return false
    }
    
    const sTable = {}
    const tTable = {}
    
    for (let char of s) {
        sTable[char] = (sTable[char] || 0) + 1
    }
    
    for (let char of t) {
        tTable[char] = (tTable[char] || 0) + 1
    }
    
    for (let char in sTable) {
        if (sTable[char] !== tTable[char]) {
            return false;
        }
    }
    
    return true;
};

File: 0242-valid-anagram/0242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        
        # base case
        if len(s) != len(t):
            return False
        
        # split and sort both strings
        # iterate and compare characters in each string
        sort_s = sorted(s)
        sort_t = sorted(t)
        
        for i in range(len(sort_s)):
            if sort_s[i] != sort_t[i]:
                return False
        
        return True
                
        